Multilingual Trait Addition

DatBoiTim opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Adds a new trait/quirk called Multilingual, allows players on spawn to understand(and speak if biologically compatible) an additional language they would not already. This trait cannot provide players with antagonist related languages, such as codespeak, narsian, or ratvarian.

## Why It's Good For The Game

With recent changes to languages such as language icons and the foreigner quirk, I figured giving players access to more languages would allow for the language system to become somewhat more prevalent.

## Changelog
:cl: DatBoiTim
add: Multilingual
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

DatBoiTim committed:
MultiLingual Trait

Multilingual Trait Created: Gives Character 1 additional non-antagonist related language(randomized). This is also designed not to overlap with languages already possessed, and has no effect on curators.

DatBoiTim committed:
Merge pull request #1 from BeeStation/master

Adds volume up/down as valid keypresses (#2226)

DatBoiTim committed:
Merge branch 'master' of https://github.com/DatBoiTim/BeeStation-Hornet

Firefox13 commented:
Fix your changelog, it's completely wrongly formatted.

yorii commented on code/datums/traits/good.dm:
```diff
@@ -96,6 +96,32 @@
+	LH.add_blocked_language(/datum/language/codespeak, LANGUAGE_ATOM)
+	LH.grant_all_languages(TRUE, TRUE, FALSE, LANGUAGE_ATOM)
+	var/datum/language/newLang
+	if(H.job != "Curator")
```
You should put all of the above procs under this "if" clause, since there is no point to do any logic if the job is curator.

DatBoiTim replied:
Fair point, I can refactor that quickly enough.

DatBoiTim replied:
Got that rearranged, now procs only run in situations they will be used.

DatBoiTim committed:
Update good.dm

PowerfulBacon commented on code/datums/traits/good.dm:
```diff
@@ -96,6 +96,33 @@
+
+	var/datum/language/newLang
+	if(H.job != "Curator")
+		LH.add_blocked_language(/datum/language/narsie, LANGUAGE_ATOM)
```
Isn't this just going to null.add_blocked_language(/datum/language/narsie, LANGUAGE_ATOM) and runtime and not work at all, or am I misunderstanding how add_blocked_language works?

DatBoiTim replied:
LH is effectively a library of nonantag languages from which a random one is pulled on spawn. Since H has no languages blocked and is not copying the language setup of LH, languages like Narsian, Codespeak, etc. can still be added when the conditions fit.

qwertyquerty replied:
Yeah but you never initialize LH

PowerfulBacon commented on code/datums/traits/good.dm:
```diff
@@ -96,6 +96,33 @@
+		LH.grant_all_languages(TRUE, TRUE, FALSE, LANGUAGE_ATOM)
+		do
+			newLang = LH.get_random_understood_language()
+		while(H.has_language(newLang, FALSE) == TRUE)
```
Would be nice to have a sanity check just in case somehow, someone has all understood languages when they gain the quirk to stop the server getting stuck in a while loop and dying

DatBoiTim replied:
That loop should only execute if the player is not a Curator(the only setup where this would be applied, to my understanding) as it is applied under that initial if(). Unless AI initialization is weird and also runs quirks through it, this shouldn't happen.

This'll probably take form of an iterative.

DatBoiTim replied:
Iterative counter added. If somehow 15 random nonantag language rolls don't yield a language not possessed yet, the loop will break. No More Infinite Loops!

DatBoiTim committed:
Update good.dm

Iterative sanity check. Count does not strictly match potential languages.

yorii commented:
Did you test it? xd

PowerfulBacon commented on code/datums/traits/good.dm:
```diff
@@ -96,6 +96,33 @@
+		do
+			newLang = LH.get_random_understood_language()
+			i = i + 1
+		while(H.has_language(newLang, FALSE) == TRUE || i < 15)
```
```suggestion
		while(H.has_language(newLang, FALSE) == TRUE && i < 15)
```

DatBoiTim committed:
Update code/datums/traits/good.dm

Co-authored-by: PowerfulBacon <jakeymortlock@gmail.com>

TeomanTheGreat commented:
good pr

qwertyquerty commented on code/datums/traits/good.dm:
```diff
@@ -96,6 +96,33 @@
+
+/datum/quirk/multilingual/on_spawn()
+	var/mob/living/carbon/human/H = quirk_holder
+	var/datum/language_holder/LH
```
Bacon was right, this appears to have never been initialized

DatBoiTim replied:
Yeah I realized that fuck up. I’ll get that changed this evening when I get off my shift. Expect a commit for that eventually.

DatBoiTim replied:
So I did some more refactoring and scrapped LH entirely.

DatBoiTim committed:
Finally Compiles and Executes

DatBoiTim committed:
Merge branch 'master' of https://github.com/DatBoiTim/BeeStation-Hornet---Personal-Fork

DatBoiTim committed:
Accounting For Tongue

Credit again to Yowii/Yorii/Yoworii for further evolving my shitcode

DatBoiTim commented:
Final iteration(I hope) of the language library used to pull a language from. Thank you Yorii!

DatBoiTim committed:
Merge pull request #3 from DatBoiTim/Experimental-Brach

Experimental brach

yorii commented:
You'll need to fix your changelog too, this will not parse.
It should look like this:

🆑 DatBoiTim
Add: Multilingual Trait
/🆑

yorii commented:
Can I be the first to suggest that it should just cost 1 point instead of 2?

PowerfulBacon commented on code/datums/traits/good.dm:
```diff
@@ -96,6 +96,28 @@
+	if(H.job != "Curator")
+		var/obj/item/organ/tongue/T = H.getorganslot(ORGAN_SLOT_TONGUE)
+		var/list/languages_possible = T.languages_possible
+		languages_possible = languages_possible - typecacheof(/datum/language/common) - typecacheof(/datum/language/uncommon) - typecacheof(/datum/language/codespeak) - typecacheof(/datum/language/narsie) - typecacheof(/datum/language/ratvar)
```
🤮

yorii replied:
Name any better method!

DatBoiTim replied:
It was either this, or to play with the LANGUAGE_MULTILINGUAL source to block those languages, grant all languages, pick a random language, remove all languages, unblock those languages, and then add the random language.

I'd take this over code getting fucky sometimes and absolutely ruining the trait 1/1000 times.

yorii replied:
Although this one still has dumb random loop 2 lines down, it's just a very astronomically small chance it would fail for more than a few loops... but in theory it could get stuck there forever 😄

PowerfulBacon commented on code/datums/traits/good.dm:
```diff
@@ -96,6 +96,28 @@
+		var/list/languages_possible = T.languages_possible
+		languages_possible = languages_possible - typecacheof(/datum/language/common) - typecacheof(/datum/language/uncommon) - typecacheof(/datum/language/codespeak) - typecacheof(/datum/language/narsie) - typecacheof(/datum/language/ratvar)
+		var/datum/language/random_language = pick(languages_possible)
+		while(random_language == H.can_speak_language(random_language))
```
use pick and take from languages_possible and then have a LAZYLEN(languages_possible) in the while loop check so that if someone knows all languages it won't kill the server instantly and will also be more efficient since it won't be checking the same language multiple times.

Or, better yet make languages possible subtract the list of languages they already know so you won't be picking a language they know and won't need a loop in the first place.

DatBoiTim replied:
Loop is now no longer present. Can only select unknown, unblocked languages that are not specific to antagonists. If statement added to ensure that a language is not picked if none are available(thus preventing exceptions when all languages are either known or blocked.

DatBoiTim committed:
Removes Loop

No more loop. Added if statement to prevent a language from being picked if no languages are available to pick.

PowerfulBacon commented on code/datums/traits/good.dm:
```diff
@@ -96,6 +96,32 @@
+		languages_possible = languages_possible - H.language_holder.understood_languages
+		languages_possible = languages_possible - H.language_holder.spoken_languages
+		languages_possible = languages_possible - H.language_holder.blocked_languages
+		if(languages_possible.len > 0)
```
```suggestion
		if(LAZYLEN(languages_possible.len))
```

LordSaladin commented:
imagine picking robo and using this to talk to borgs/AI.

DatBoiTim commented:
You’d need a compatible tongue for that, at the very least

DatBoiTim committed:
Now Using LAZYLEN

Does what it says on the tin

PowerfulBacon commented:
Code is good but damn merge conflict

DatBoiTim committed:
Merge branch 'master' into pr/4

DatBoiTim committed:
Merge pull request #5 from DatBoiTim/pr/4

Pr/4

PowerfulBacon commented on code/datums/traits/good.dm:
```diff
@@ -96,6 +96,32 @@
+		else
+			return
+	else
+		return
```
```suggestion
```

PowerfulBacon replied:
These returns are pointless because they are at the end of the proc anyway.

DatBoiTim replied:
Removed - Tim

DatBoiTim committed:
Removed unecessary Else Return Statements
